### PR TITLE
Added tooltip

### PR DESCRIPTION
**Pull request by lorem--ipsum**
Pouet

### CODE REVIEW

File: src/LineChart.ts [hunk 1]
```diff
@@ -12,7 +14,9 @@ module n3Charts {
     public scope = {
       data: '=',
       options: '=',
-      styles: '='
+      styles: '=',
+      onDatumOver: '=',
+      onDatumOut: '='
```
Comment (chaosmail) [inline src/LineChart.ts]: sweet

File: src/factories/Tooltip.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
+module n3Charts.Factory {
+  'use strict';
+
+  export class Tooltip extends Utils.BaseFactory {
+
+    private tooltip:D3.Selection;
```
Comment (chaosmail) [inline src/factories/Tooltip.ts]: can we call this `svg`? All other factory containers are called like this

File: src/factories/Tooltip.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        .style('display', 'inline-block');
+
+      tooltip.append('div')
+        .attr('class', 'value')
+        .style('display', 'inline-block');
+    }
+
+    destroy() {
+      this.eventMgr.on('over.tooltip', null);
+      this.eventMgr.on('out.tooltip', null);
```
Comment (chaosmail) [inline src/factories/Tooltip.ts]: great, the first time we use `destroy` for what it was meant to be :)

File: src/factories/Tooltip.ts (hunk 2 above)
Comment (lorem--ipsum) [inline src/factories/Tooltip.ts]: Yep, good point

File: src/factories/series/Column.ts [hunk 4]
```diff
@@ -9,50 +9,66 @@ module n3Charts.Factory.Series {
     protected seriesClass: string = Column.type + '-series';
     protected dataClass: string = Column.type;
 
-    public colsWidth: number = 0;
+    private gapFactor: number = 0.2;
+    private outerPadding: number = (this.gapFactor / 2) * 3;
+    private columnsWidth: number = 0;
+
+    public x1Scale: D3.Scale.OrdinalScale;
```
Comment (chaosmail) [inline src/factories/series/Column.ts]: can we call this `xOrdinalScale` or sth?

File: src/factories/series/Column.ts [hunk 5]
```diff
@@ -9,50 +9,66 @@ module n3Charts.Factory.Series {
...
 
     update(data: Utils.Data, options: Utils.Options) {
       super.update(data, options);
 
       var series = options.getSeriesForType(Column.type);
 
-      this.updateColsWidth(series);
+      this.updateColumnsWidth(series, options);
```
Comment (chaosmail) [inline src/factories/series/Column.ts]: In `SeriesFactory` we can access the options with `this.options`. Is there a good reason not to do so?

File: src/factories/series/Column.ts (hunk 4 above)
Comment (lorem--ipsum) [inline src/factories/series/Column.ts]: Hmmmmm `innerScale` ?

File: src/factories/series/Column.ts [hunk 6]
```diff
@@ -9,50 +9,66 @@ module n3Charts.Factory.Series {
...
+      this.columnsWidth = delta < Number.MAX_VALUE ? delta / series.length : 10;
     }
 
-    updateData(group: D3.Selection, series: Utils.Series, index: number, numSeries: number) {
+    updateColumnScale(series: Utils.Series[], options: Utils.Options) {
+      var halfWidth = this.columnsWidth * series.length / 2;
+
+      this.x1Scale = d3.scale.ordinal()
+        .domain(series.map((s) => s.id))
+        .rangeBands([-halfWidth, halfWidth], 0, 0.1);
```
Comment (chaosmail) [inline src/factories/series/Column.ts]: awesome, finally we use some built-in stuff :D

File: src/factories/series/Column.ts (hunk 5 above)
Comment (lorem--ipsum) [inline src/factories/series/Column.ts]: Yeah, I don't like the idea of having semi-global stuff... They're given through the parameters anyway, and having function that don't depend on outer state is better IMHO.

File: src/factories/series/Column.ts (hunk 6 above)
Comment (lorem--ipsum) [inline src/factories/series/Column.ts]: Yeah, it just took me hours to figure out how to use this scale ^^

File: src/factories/series/Column.ts [hunk 7]
```diff
@@ -9,50 +9,66 @@ module n3Charts.Factory.Series {
...
       };
 
       var updateCol = (s) => {
         s.attr({
-          x: (d) => xAxis.scale(d.x) - (numSeries * 0.5 - index) * this.colsWidth,
-          y: (d) => yAxis.scale(d.y),
-          width: this.colsWidth * (1 - colsWidthSpacing),
-          height: (d) => yAxis.scale(0) - yAxis.scale(d.y)
+          x: xFn,
+          y: (d) => d.y > 0 ? yAxis.scale(d.y) : yAxis.scale(0),
```
Comment (chaosmail) [inline src/factories/series/Column.ts]: we could also write `yAxis.scale(d3.min(0, d.y))`

File: src/utils/Options.ts [hunk 8]
```diff
@@ -34,7 +34,11 @@ module n3Charts.Utils {
     }
 
     getAbsKey(): string {
-      return this.axes.x.key;
+      if (!!this.axes.x) {
```
Comment (chaosmail) [inline src/utils/Options.ts]: do we need `!!` here or can we remove them?

File: test/e2e/templates/column.tpl.html [hunk 9]
```diff
@@ -42,4 +52,4 @@
     {x: 7, val_0: 4.927, val_1: 3.35, val_2: -13.074, val_3: -12.625}
   ]
 }
-{% endblock %}
```
Comment (chaosmail) [inline test/e2e/templates/column.tpl.html]: are you using a plugin for this?

File: test/spec/testUtils.spec.ts [hunk 10]
```diff
@@ -0,0 +1,169 @@
+var utils;
+
+// This needs to be cleaned, and maybe externalized into a small testing
+// framework utils. But then we'd need to find a name for it.
+utils = angular.module('testUtils', []);
```
Comment (chaosmail) [inline test/spec/testUtils.spec.ts]: alright, good you added it - there is some very useful stuff included. however, a lot of things (dealing with mouse, classes, dom elements, ..) should be now supported with protractor

File: src/factories/series/Column.ts (hunk 4 above)
Comment (chaosmail) [inline src/factories/series/Column.ts]: perfect

File: src/factories/series/Column.ts (hunk 5 above)
Comment (chaosmail) [inline src/factories/series/Column.ts]: yeah i agree on that, we should then find a solution that we can apply everywhere

**Comment by chaosmail** [comment]
very nice PR, I am amazed where this is going.

**Comment by chaosmail** [comment]
maybe we can agree on some naming conventions for the branches, like `dev/feat-tooltip`

**Comment by lorem--ipsum** [comment]
Okay, I'll take care of this tomorrow or Monday (flying tomorrow, so I don't know if I'll have time) :-)

Thanks for your remarks !

File: src/factories/series/Column.ts (hunk 7 above)
Comment (lorem--ipsum) [inline src/factories/series/Column.ts]: The linter doesn't allow it for some reason...

File: test/e2e/templates/column.tpl.html (hunk 9 above)
Comment (lorem--ipsum) [inline test/e2e/templates/column.tpl.html]: what do you mean ?

File: test/spec/testUtils.spec.ts (hunk 10 above)
Comment (lorem--ipsum) [inline test/spec/testUtils.spec.ts]: Protractor's way of dealing with events is pretty limited IMHO... It doesn't handle mouseover/mouseout events, for example

File: test/e2e/templates/column.tpl.html (hunk 9 above)
Comment (chaosmail) [inline test/e2e/templates/column.tpl.html]: Sorry my question looks weird in this context; it seems you are automatically appending line-breaks to the end of files where there is none - which is nice. So I thought you have a plugin that does this automatically :) (so I can install it also)

**Pull request merged**